File: backend/src/main/java/outcobra/server/model/Subject.java
package outcobra.server.model;

import java.util.List;

import javax.persistence.*;
import javax.validation.constraints.NotNull;

import lombok.*;

@Data
@Entity
@AllArgsConstructor
@NoArgsConstructor
public class Subject {
    @Id
    private Long id;

    @NotNull
    private String name;

    @ManyToOne
    private Semester semester;

    @OneToMany(mappedBy = "subject")
    private List<TimetableEntry> timetableEntries;

    @OneToMany(mappedBy = "subject")
    private List<Task> tasks;

    @OneToMany(mappedBy = "subject")
    private List<MarkReportEntry> markReportEntries;

    @OneToMany(mappedBy = "subject")
    private List<Exam> exams;

    @OneToOne
    private MarkGroup markGroup;

    @ManyToOne
    private Teacher teacher;
}

File: backend/src/main/kotlin/outcobra/server/model/Subject.kt
package outcobra.server.model

import com.querydsl.core.annotations.QueryInit
import outcobra.server.model.interfaces.ParentLinked
import javax.persistence.*
import javax.validation.constraints.NotNull

@Entity
data class Subject(@NotNull var name: String = "",
                   @NotNull @Enumerated(EnumType.STRING) var color: Color? = null,
                   @ManyToOne @QueryInit("schoolYear.schoolClass.institution.user") var semester: Semester? = null,
                   @OneToMany(mappedBy = "subject") var timetableEntries: List<TimetableEntry> = listOf(),
                   @OneToMany(mappedBy = "subject") var tasks: List<Task> = listOf(),
                   @OneToMany(mappedBy = "subject") var markReportEntries: List<MarkReportEntry> = listOf(),
                   @OneToMany(mappedBy = "subject") var exams: List<Exam> = listOf(),
                   @OneToOne var markGroup: MarkGroup? = null,
                   @ManyToOne var teacher: Teacher? = null)
    : ParentLinked, AbstractEntity() {

    override val parent: ParentLinked?
        get() = semester
}
